pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,317 @@
+	processExcluder *process.Excluder
+	gvkAggregator   *aggregator.GVKAgreggator
+	// mu guards access to any part of the spec above
+	mu sync.RWMutex
+
+	opa              syncutil.OpaDataClient
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	registrar        *watch.Registrar
+	watchedSet       *watch.Set
+	replayErrChan    chan error
+	replayTicker     time.Ticker
+	reader           client.Reader
+	excluderChanged  bool
+}
+
+func NewCacheManager(config *CacheManagerConfig) (*CacheManager, error) {
+	if config.WatchedSet == nil {
+		return nil, fmt.Errorf("watchedSet must be non-nil")
+	}
+	if config.Registrar == nil {
+		return nil, fmt.Errorf("registrar must be non-nil")
+	}
+	if config.ProcessExcluder == nil {
+		return nil, fmt.Errorf("processExcluder must be non-nil")
+	}
+	if config.Tracker == nil {
+		return nil, fmt.Errorf("tracker must be non-nil")
+	}
+	if config.Reader == nil {
+		return nil, fmt.Errorf("reader must be non-nil")
+	}
+
+	cm := &CacheManager{
+		opa:              config.Opa,
+		syncMetricsCache: config.SyncMetricsCache,
+		tracker:          config.Tracker,
+		processExcluder:  config.ProcessExcluder,
+		registrar:        config.Registrar,
+		watchedSet:       config.WatchedSet,
+		reader:           config.Reader,
+	}
+
+	cm.gvkAggregator = aggregator.NewGVKAggregator()
+
+	cm.replayTicker = *time.NewTicker(3 * time.Second)
+
+	return cm, nil
+}
+
+func (c *CacheManager) Start(ctx context.Context) error {
+	go c.updateDatastore(ctx)
+
+	<-ctx.Done()
+	return nil
+}
+
+// AddSource adjusts the watched set of gvks according to the newGVKs passed in
+// for a given {syncSourceType, syncSourceName}.
+func (c *CacheManager) AddSource(ctx context.Context, newGVKs []schema.GroupVersionKind, syncSourceType, syncSourceName string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	opKey := aggregator.Key{Source: syncSourceType, ID: syncSourceName}
+
+	if err := c.gvkAggregator.Upsert(opKey, newGVKs); err != nil {
+		return fmt.Errorf("internal error adding source: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CacheManager) RemoveSource(ctx context.Context, syncSourceType, syncSourceName string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if err := c.gvkAggregator.Remove(aggregator.Key{Source: syncSourceType, ID: syncSourceName}); err != nil {
+		return fmt.Errorf("internal error removing source: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CacheManager) ExcludeProcesses(newExcluder *process.Excluder) {
+	if c.processExcluder.Equals(newExcluder) {
+		return
+	}
+
+	c.mu.Lock()
+	c.processExcluder.Replace(newExcluder)
+	// there is a new excluder which means we need to schedule a wipe for any
+	// previously watched GVKs to be re-added to get a chance to be evaluated
+	// for this new process excluder.
+	c.excluderChanged = true
+	c.mu.Unlock()
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
```

Comment:
I believe we should also guard the addition of an object with a check like:

"Is this object of a GVK that is supposed to be in the cache?"

The scenario we'd want to avoid is that a reconcile event for an object makes its way to the sync_controller after a long delay, and that object's GVK has already been removed from the sync configuration and the cache has been wiped.

This delayed event could then potentially add an object to the cache that should not be there.

Reply:
we get this "for free" with the `FilteredOpaDataClient` that we init the cache manager with.

https://github.com/open-policy-agent/gatekeeper/blob/1878de1ce150142789cccb4be9ae30c44e3ad615/pkg/syncutil/opadataclient.go#L45-L56

Reply:
Is this filtered client used elsewhere?  What is the value of this extra layer of abstraction?

Reply:
Note this was pre-existing:

https://github.com/open-policy-agent/gatekeeper/blob/b2def12693241c7018cdb9fa1b55e62919d7d782/pkg/controller/sync/opadataclient.go#L32-L43

used in the config_controller before: 

https://github.com/open-policy-agent/gatekeeper/blob/b2def12693241c7018cdb9fa1b55e62919d7d782/pkg/controller/config/config_controller.go#L112-L122

> What is the value of this extra layer of abstraction?

AFAIU, this solves for 

> "Is this object of a GVK that is supposed to be in the cache?"

I'd vote not to change the tool for that now but am open to revisit the question of whether we want to do it this way if that make sense.

Reply:
FilteredClient was intended to solve this issue. WatchManager is a better entity to own that logic now, since we are aiming to centralize concerns like this and make it easier to verify they are being addressed.

Reply:
Thanks all for the feedback. CacheManager now is the FilteredClient. Tests added for it too.